src/gc.c

```diff
@@ -992,6 +993,8 @@ STATIC_INLINE jl_value_t *jl_gc_big_alloc_inner(jl_ptls_t ptls, size_t sz)
         jl_atomic_load_relaxed(&ptls->gc_num.allocd) + allocsz);
     jl_atomic_store_relaxed(&ptls->gc_num.bigalloc,
         jl_atomic_load_relaxed(&ptls->gc_num.bigalloc) + 1);
+    jl_atomic_fetch_add_relaxed(&gc_heap_stats.bytes_mallocd, allocsz);
+    jl_atomic_fetch_add_relaxed(&gc_heap_stats.heap_size, allocsz);
```

Comment:
It is probably not acceptable to add atomic seq-cst instructions (there were none before, on Intel, since they will slow this function down unacceptably, especially since these are memory conflicts.

Reply:
Are these `seq-cst`?, you mean they being `fetch_add` instead of a load + store?

Reply:
The issue is that we have to have an update to the global variable somewhere and I think it needs to be this otherwise it will be incorrect?

Reply:
One idea is to have a thread local value and keep that around and batch update it?

Reply:
The add to the thread local is pretty easy, the more complicated thing might be when to do the batch update?

Reply:
Interestingly, the worse decisions caused by the batching seem to have a bigger impact than the atomics themselves, but they are a cost in some applications for sure.

Reply:
Do any of you have any ideas?

Reply:
(If a thread still has an allocation chunk, then it simply allocates by decrementing it until it needs to request a new chunk)

Reply:
That's an interesting idea, though it seems to be just a different (possibly) cleaner way of implementing it.

Reply:
Most of these should be global with atomics, but probably not the individual malloc/free calls

Reply:
I implemented batching but wasn't satisfied with it because in some scenarios (low live bytes) it made worse decisions.

Reply:
One algorithm to try would be a lockless binary sum. Here's the sketch of the algorithm
```
n = nextpow(2, nthreads)
big_bytes = zeros(n*log2(n)) # big_bytes[i] stores big_bytes[2*i] + big_bytes[2*i+1]
function malloced_bytes(thread_i, n_bytes)
    ind = (n-1)*log2(n) + thread_i
    big_bytes[ind] += n_bytes
    for j in 1:log2(n)
         ind >>= 1
         # no atomic needed. This can be racy and still very close to accurate
         big_bytes[ind] =  big_bytes[2*ind] + big_bytes[2*ind+1]
   end
   return big_bytes[1]
end
```
This will give a very close estimate at all times since it gives an exact allocation count modulo race conditions (and with race conditions, it will miss a maximum of ~1 allocation per thread and in reality, I think it will be a lot less). I'm not 100% sure the loop will be faster than the atomic, but it is a pretty simple loop.

Reply:
It also probably would be better to do this with a 4 or 8 wide tree since the loop would be half as long (with more additions per loop)

Reply:
So the issue with this might be that foreign threads are a thing that exists.

Reply:
why is that a problem?

Reply:
So I was looking a bit further, we fully disable the GC when adding the threads so I think it's fine in the end. I am implementing something that uses your idea.